src/features/device-logs/lib/backends/loki.ts

```diff
@@ -328,21 +323,18 @@ export class LokiBackend implements DeviceLogsBackend {
 		ctx: LogWriteContext,
 		logs: Array<DeviceLog & { version?: number }>,
 	) {
-		let nonce = 0;
 		const streams: StreamAdapter[] = [];
 		const streamIndex: { [key: string]: StreamAdapter } = {}; // index streams by labels for fast lookup
 		for (const log of logs) {
 			this.validateLog(log);
 			log.version = VERSION;
-			const [seconds, nanoseconds] = convertToNanoseconds(
-				log.timestamp,
-				nonce++,
-			);
 			const timestamp = new Timestamp();
-			timestamp.setSeconds(seconds);
-			timestamp.setNanos(nanoseconds);
+			timestamp.setSeconds(Math.floor(Number(log.nanoTimestamp / 1000000000n)));
+			timestamp.setNanos(Number(log.nanoTimestamp % 1000000000n));
 			// store log line as JSON
-			const logJson = JSON.stringify(log);
+			const logJson = JSON.stringify(log, (key, value) =>
+				key === 'nanoTimestamp' ? undefined : value,
+			);
```

Comment:
Do you know how this performs vs
```ts
delete log.nanoTimestamp
const logJson = JSON.stringify(log);
```
by any chance?

Reply:
Here's a benchmark

```
const loadNs = BigInt(Date.now()) * 1000000n - process.hrtime.bigint();

function getNanoTimestamp() {
	return loadNs + process.hrtime.bigint();
}

const createLog = (extra = {}) => {
	return {
		isStdErr: true,
		isSystem: true,
		message: `a log line`,
		nanoTimestamp: getNanoTimestamp(),
		timestamp: Date.now(),
		createdAt: Date.now(),
		...extra,
	};
};

function genLogs(removeNanos = false) {
	const logs = []
	for (let i = 0; i < 1e6; i++) {
		const log = createLog()
		if (removeNanos) {
			// @ts-ignore
			delete log.nanoTimestamp
		}
		logs.push(log)
	}
	return logs
}

let logs = genLogs()

let startTime = Date.now()
for (const log of logs) {
	// @ts-ignore
	delete log.nanoTimestamp
	JSON.stringify(log)
}
console.log('Delete took', Date.now() - startTime);

logs = genLogs()
startTime = Date.now()
for (const log of logs) {
	JSON.stringify(log, (key, value) =>
		key === 'nanoTimestamp' ? undefined : value)
}
console.log('Replacer took', Date.now() - startTime);

logs = genLogs(true)
startTime = Date.now()
for (const log of logs) {
	JSON.stringify(log)
}
console.log('Noop took', Date.now() - startTime);
```

Surprisingly delete is slower 
```
Delete took 2131
Replacer took 1664
Noop took 1141
```

Reply:
Yeah it's probably due to how it interacts with the hidden classes and fast props stuff, in that case I'm fine as-is